[1단계 - 출석 구현] 에드(김진우) 미션 제출합니다.

jinu0328 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [Prolog 링크](https://prolog.techcourse.co.kr/studylogs)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
도메인에서의 getter, setter를 일부 허용하였기 때문입니다..

## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

## 1. 도메인 설계 및 테스트 전략 관련 질문

 제가 TDD가 처음이라, 처음부터 도메인을 어느 정도 설계하고 테스트를 진행해야 하는지 잘 모르겠습니다.

기능을 구현할 때, 처음부터 도메인을 나누고 해당 도메인 내에서 테스트를 진행하는 것이 적절한 지 하나의 클래스에서 기능을 먼저 구현하고, 한 기능이 완성된 후 도메인을 분리하는 것이 더 적절한 지 아직 판단이 잘 서지 않은 것 같습니다. (이번 미션 수행에 있어서는 후자의 방식을 선택하였습니다.) 이와 관련해서 이프의 조언을 들어보고싶습니다!

## **2. `List<Attendance>`에 대한 설계 고민 질문**

해당 미션에서 `Crew`라는 도메인이 `List<Attendance>`를 멤버로 가지고 출석 기록을 추가/수정하는 방식으로 구현했습니다.

이 과정에서 `Attendance`의 일급 컬렉션(`Attendances`)을 따로 두고 이를 `Crew`의 멤버로 둘지,아니면 `List<Attendance>`를 직접 멤버로 둘지에 대한 고민이 있었습니다.

제 생각을 정리한 각 접근 방식의 장단점은 다음과 같습니다.

### 1. `Attendances` (일급 컬렉션) 사용

장점

- 비즈니스에 종속적인 컬렉션을 구현하여 가독성이 증가→ `Attendances`가 출석 목록을 관리하는 역할을 맡아, `Crew`가 컬렉션을 직접 다루지 않아도 됨.
- 여러 `Attendance`에 걸친 상태와 행위를 한 곳에서 관리 가능→ `Attendances` 내부에서 출석 관련 로직을 관리할 수 있어 응집도가 높아짐.

단점

- `Attendance`는 계속해서 추가/수정이 필요한 객체인데, 일급 컬렉션의 목적인 불변성이 보장되지 않음.
- 불변성이 깨진 일급 컬렉션은 믿고 사용할 수 없을 것 같다.→ `Attendances`가 내부적으로 `List<Attendance>`를 가지고 있어도, 결국 `add()`나 `update()`가 필요한 상황이 발생 → 일급 컬렉션의 불변성 원칙이 깨지는 문제


### 2. `List<Attendance>`를 직접 사용

장점

- 출석 목록이 계속 추가/수정되는 경우, 리스트를 직접 관리하는 것이 자연스럽다.→ 불변성을 깨트리는 위험 없이 컬렉션을 직접 수정 가능.

단점

- 여러 `Attendance`에 대한 상태/행위를 `Crew` 도메인에서 책임지게 됨.→ `Crew`가 출석 관리 로직을 직접 다루게 되면서 역할이 많아짐.→ `Crew`가 `List<Attendance>`를 다루는 여러 메서드를 가지게 되면, 응집도가 떨어짐


현재 미션에서 2번을 선택한 이유는 출석 기록(`Attendance`)이 계속 추가/수정되어야 하는 요구사항이 있기 때문에, 불변성이 중요한 일급 컬렉션의 개념과 잘 맞지 않는다고 생각했기때문입니다.

하지만, 여러 `Attendance`를 다루는 상태와 행위를 `Crew`가 관리하는 것이 적절한지,아니면 이를 `Attendances`라는 별도의 도메인으로 분리하는 것이 더 나은지 판단이 어렵습니다.

이러한 상황에서, 불변성이 깨지는 일급 컬렉션을 사용해도 괜찮을까요? 아니면 `Crew`에서 직접 `List<Attendance>`를 관리하는 것이 더 적절한 접근일까요?

---

## **3. Day 객체를 만든 이유 및 유지 여부에 대한 고민**

현재 `Day` 객체는 사실상 `LocalDate`를 한 번 감싼 정도의 역할을 하고 있습니다.

`Day` 객체를 도입한 주된 이유는 주말을 제외한 공휴일 처리를 위한 것이었으며,

일반적인 방법으로 생각했을 때 공휴일을 직접 관리하기 위해 `Day` 객체를 활용하는 것이 적절하다고 판단했습니다.

하지만, 현재 미션에서 사용하는 CSV 파일은 2월을 기준으로 작성되었고, 2월에는 공휴일이 없기 때문에 사실상 `Day` 객체가 하는 역할이 거의 없습니다.

추가적으로 날짜를 `Day`라는 커스텀 클래스 객체로 관리하게 되면서 도메인간의 계층 구조가 더욱 복잡해진 것도 사실입니다.

이러한 상황에서,

1. 향후 확장을 고려하여 `Day` 객체를 그대로 유지할지
2. 현재 필요하지 않으므로 `Day` 객체를 제거하고 `LocalDate`를 직접 사용할지에 대한 고민이 있습니다.

horizonpioneer committed:
docs(readme): 1차 요구사항 정리

horizonpioneer committed:
test: 출석 확인 정상 출력 테스트 구현

horizonpioneer committed:
feat(attendanceCheck): 등교 시간에 따라 출석 상태를 체크하는 기능 구현

horizonpioneer committed:
test: 등교 시간에 따른 출석 상태 변동 테스트 구현

horizonpioneer committed:
feat(attendanceCheck): 날짜에 따라 해당 요일을 반환하는 기능 구현

horizonpioneer committed:
test: 출석 확인 시, 현재 날짜가 출력되는 테스트 구현

horizonpioneer committed:
feat(AttendanceCheck): 문자열을 LocalTime으로 변환하는 기능 구현

horizonpioneer committed:
test: 문자열을 LocalTime으로 변환하는 테스트 구현

horizonpioneer committed:
feat(AttendanceCheck): 날짜에 따른 등교 기준 시간을 반환하는 기능 구현

horizonpioneer committed:
test: 날짜에 따른 출석 기준 시간 적용 테스트 구현

horizonpioneer committed:
refactor: InputConverter 유틸 클래스 구현

- 문자열을 LocalTime으로 변환하는 메서드를 분리

horizonpioneer committed:
refactor: 요일 및 출석 시간 enum화한 DayOfWeek 구현

- id를 기반으로 요일 정보를 반환하는 정적 메서드 구현

horizonpioneer committed:
refactor: Day와 Attendance 도메인 분리

- 출석 상태를 확인하는 기능 분리

horizonpioneer committed:
refactor: AttendanceDto 구현

horizonpioneer committed:
refactor: LocalDate를 기반으로 DayOfWeek 객체를 반환하는 기능 구현

horizonpioneer committed:
refactor: Dto 기반으로 테스트 수정

horizonpioneer committed:
test: 수정시간에 따라 기존 출석시간 변경 테스트 구현

horizonpioneer committed:
feat: AttendanceTime 변경 메서드 및 Dto 필드 추가

- setAttendanceTime을 통해 출석 시간을 변경하고 updateStatus를 호출하여 출석 상태를 변경할 수 있도록 한다.

horizonpioneer committed:
refactor: 여러 개의 출석 상태 변경 테스트를 ParameterizedTest로 통합

horizonpioneer committed:
test: 각 출석 상태별 횟수 계산 테스트 구현

horizonpioneer committed:
test: 제적 위험자 여부 판단 테스트 구현

- PenaltyStatus enum을 통해 제적 위험자 기준에 따라 StatusName 반환 기능 추가

horizonpioneer committed:
refactor: 공휴일 여부를 나타내는 필드를 메서드에서 산출하도록 변경

- isHoliday 필드 삭제 후, checkHoliday 메서드를 통해 공휴일 여부 판단

horizonpioneer committed:
refactor: 제적 상태 이름을 반환하는 기능을 제적 상태 enum을 반환하는 기능으로 수정

horizonpioneer committed:
feat: Crew 도메인 추가

horizonpioneer committed:
feat: CrewDto 클래스 구현

horizonpioneer committed:
feat: CrewDto의 중첩 Dto CrewDtos 구현

horizonpioneer committed:
env: 출석 기록 파일 추가

horizonpioneer committed:
feat: 출석 기록으로부터 크루원 초기 데이터 세팅

- 출석 기록을 csv 파일에서 가져오는 기능 구현

horizonpioneer committed:
feat: LocalTime을 String으로 변환하는 기능 구현

horizonpioneer committed:
feat: 기존 출석 기록이 있는지 확인하는 메서드 구현

horizonpioneer committed:
fix: csv 파일 경로 수정

horizonpioneer committed:
feat: InputView 클래스 구현

horizonpioneer committed:
feat: OutputView 클래스 구현

horizonpioneer committed:
feat: 닉네임을 통해 Crew를 반환하는 메서드 구현

horizonpioneer committed:
refactor: AttendanceDto 멤버로 출석 일자 추가

horizonpioneer committed:
feat: 문자열을 정수로 바꾸는 정적 메서드 구현

horizonpioneer committed:
chore: 상수명 변경

horizonpioneer committed:
feat: 일자에 따라 Attendance 객체를 반환하는 기능 구현

horizonpioneer committed:
feat: Crew의 출석 기록을 반환하는 기능 구현

- 날짜 기반 정렬 후, 방어적 복사를 이용하여 getter 구현

horizonpioneer committed:
feat: 출석하지 않은 날짜를 일괄적으로 결석 처리하는 기능 구현

horizonpioneer committed:
refactor: 분리된 도메인으로 테스트 재구성

- 임시 클래스인 AttendanceCheck 삭제 후, 테스트 수정

horizonpioneer committed:
feat: OutputView 클래스 구현

horizonpioneer committed:
refactor: covertLocalTimeToString 메서드에 null 처리 추가

horizonpioneer committed:
refactor: 공휴일에 Attendance 객체가 생성되는 경우, 예외 처리 추가

horizonpioneer committed:
refactor: crewDtos 정렬 방식 수정

horizonpioneer committed:
refactor: 출석 상태 확인 메서드에 null 처리 구현

horizonpioneer committed:
feat: CrewDto의 리스트 생성 메서드 구현

horizonpioneer committed:
refactor: AttendanceHistoryLoader 메서드 분리

horizonpioneer committed:
feat: AttendanceController 구현

horizonpioneer committed:
refactor: AttendanceController 메서드 분리

horizonpioneer committed:
feat: Application 추가

horizonpioneer committed:
refactor: Day 멤버 final로 선언

horizonpioneer committed:
chore: getStandardTimeById 메서드 삭제

horizonpioneer committed:
refactor: DayofWeek 필드 제거

- 내부 메서드를 통해 DayOfWeek 조회하도록 수정

sinb57 commented on README.md:
```diff
@@ -1,3 +1,142 @@
+
+# 입력
+
+- [ ]  실행할 기능을 숫자로 입력 받는다
+    - [ ]  1 → 출석 확인
+    - [ ]  2 → 출석 수정
+    - [ ]  3 → 크루별 출석 기록 확인
+    - [ ]  4 → 제적 위험자 확인
+    - [ ]  Q를 입력하면 시스템을 종료한다
+        - [ ]  1, 2, 3, 4, Q가 아닌 문자가 들어왔을 때 예외처리
```
요구사항 구현을 마쳤다면 체크 부탁드려요~

jinu0328 replied:
기능이 완성될 때 마다 README의 체크리스트도 함께 수정했어야됐는데 미처 신경쓰지 못했던 것 같습니다.. 수정하겠습니다!

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+package config;
```
config 패키지는 어떠한 역할을 지니고 있나요?

jinu0328 replied:
`config` 패키지는 보통 환경 설정, 애플리케이션의 전역적인 설정 등의 역할을 지닌다고 생각합니다. 현재 `config` 패키지에 위치한 `AttendanceHistoryLoader` 클래스는 초기 데이터를 로드하는 역할을 하기에, 이를 "전역적인 설정"의 일부로 보고 해당 패키지에 두었습니다.

하지만 다시 생각해보니, 이 클래스는 설정이라기보다는 "데이터 로드"라는 명확한 역할을 수행하고 있어 "설정"이라는 키워드와는 거리가 있는 것처럼 느껴집니다.

이런 경우, 별도의 loader 패키지로 분류하는 것이 더 직관적이고 적절할까요?

sinb57 replied:
domain 패키지에 둬도 무방하다고 생각해요. domain 패키지는 무슨 책임을 지니고 있나요? 분류하고자 하는 loader는 무슨 책임을 지니고 있나요? 다른 한편으로 생각해보면, load라는 행위는 입출력과 무관하나요?

jinu0328 replied:
`domain` 패키지는 애플리케이션의 핵심 비지니스 로직을 다룬다고 생각합니다. 현재 `loader`는 csv 형태로 저장된 데이터를 읽어 도메인 객체를 생성하는 책임을 지니고있습니다.`load`라는 행위 자체는 어쩌면 입출력과 관련된 작업이라고 생각합니다. 다만 그 결과물이 도메인 객체를 생성하고 있다는 점에서 비지니스 규칙과 도메인을 구성하는데에도 직접적으로 연결되어있다고 생각됩니다. 이프의 말대로 `domain `패키지로 분류하여도 합리적인 것 같습니다!

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+    private final DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+    private final DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HH:mm");
+
+    public Crews loadCrews() {
```
해당 로직은 테스트가 필요없을까요?

jinu0328 replied:
검증된 라이브러리를 사용하지 않고 직접 전처리 과정을 통해 파일 내용을 파싱하고 객체를 생성하는 로직이기 때문에, 테스트가 필요하다고 생각됩니다!

도메인을 분리한 이후에 새롭게 구현된 클래스에 대한 테스트 코드 추가를 놓쳤던 것 같습니다.
추가적인 테스트 코드를 작성하여 보완하겠습니다

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+    private final DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HH:mm");
+
+    public Crews loadCrews() {
+        List<Crew> crews = new ArrayList<>();
+        Map<String, Crew> crewMap = new HashMap<>();
```
crew를 등록하는 과정을 Crews 에 위임하면 Loader 로직가 가벼워질 듯 해요.

jinu0328 replied:
`Crews`라는 일급 컬렉션에 `addCrew()` 같은 메서드를 두면 불변성이 깨질 수 있다고 생각하여, 현재 방식으로 구현했습니다.

출석 애플리케이션에서는 초기 데이터 로드 이후 Crew 객체가 추가될 일이 없기 때문에, `addCrew()`를 두지 않는 것이 더 적절하다고 판단했는데요.

하지만 말씀해주신 대로, Loader에서 Crew 추가 로직을 Crews로 위임하면 로직이 더 간결해질 수 있을 것 같습니다.

결국,

`addCrew()`를 추가하여 `Loader`의 로직을 간결하게 하는 대신 `Crews`의 가변성을 허용하는 것
`Loader`에서 `Crew` 추가 로직을 유지하여 `Crews`의 불변성을 보장하는 것
이 두 가지 선택지 사이에서 고민이 됩니다! 🤔

sinb57 replied:
일급 컬렉션이 항상 불변성을 보장해야 하나요?

sinb57 replied:
https://github.com/woowacourse/java-attendance/pull/49#issuecomment-2676838680
> 제가 일급 컬렉션을 적용하지 않은 이유는 다음과 같습니다.
> - 일급 컬렉션에 대한 일반적인 기대와의 불일치
> - 불필요한 리소스 소모 방지

위의 답변을 주셨는데, 결국 일급 컬렉션은 불변성을 보장해야 한다는 내용이네요. 일급 컬렉션이 불변성을 보장해야 하나요? 

어떠한 개념에 대해 받아들일 때는 등장한 이유와 배경을 이해하는 것이 중요하고 어떠한 성질을 지니고 있는지 등 원초적인 조사가 필요하다고 생각해요. 일급 컬렉션에서 "일급"은 무슨 의미일까요? 어찌보면 이 때문에 불변성을 보장해야 한다는 내용이 탄생했을지도 몰라요. 일급 컬렉션을 처음 다루는 The ThoughtWorks Anthology(소트웍스 엔솔로지)을 찾아보시는 것도 좋겠어요.

저는 해당 케이스의 경우 필드를 변경가능토록 열어놔도 좋다고 생각해요. 만약 불변을 무조건 보장해야 한다면, Crews 객체를 다시 살펴보길 바래요. 불변성을 보장하는 것은 무슨 의미일까요? Crews는 불변을 보장하고 있나요?

jinu0328 replied:
학습 결과 일급 컬렉션의 "일급"은 "일급 객체"에서 유래하여 컬렉션을 단순히 자료구조로 사용하는 것이 아닌 **하나의 독립적인 객체로서의 책임을 갖게한다**는 의미를 가지고 있음을 알게되었습니다. 처음 일급컬렉션을 접했을 때, "컬렉션을 한번 더 포장해줌으로써 참조에 의한 값 변화를 저지한다" 라는 것이 일급컬렉션의 주요 개념이라고 이해했지만, 이프의 피드백을 받고 고민해보니 **컬렉션이 독립적인 객체로서 책임을 갖도록 만든다**라는 것이 더 중요한 개념이라는 것을 알게 되었습니다. 즉, 일급 컬렉션 또한 일반적인 객체처럼 불변성을 지향하는 것이 이상적이지만, 요구사항에 따라 반드시 불변성을 유지해야 하는 것은 아니라는 점을 배울 수 있었습니다.

실제로 Attendances를 도입하여 리팩토링 해보니 Crew가 담당하던 Attendance 관련 로직이 Attendances로 이동하면서 응집도가 높아지고 관리가 용이하다는 장점을 느낄 수 있었습니다.

 + Crews 객체를 살펴보니

```java
    public void recordAllAbsence() {
        crews.forEach(Crew::recordAbsence);
    }
```
이미 위의 메서드에서 불변성을 깨트리고있었군요..! 꼭 불변성을 유지하지 않더라도 특정 객체에 대한 행동을 한 곳에서 정의하고 관리하는 것이 충분히 이점을 제공한다는 것을 느낄 수 있었습니다.

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+        Map<String, Crew> crewMap = new HashMap<>();
+
+        try (BufferedReader reader = new BufferedReader(new FileReader("src/main/resources/attendances.csv"))) {
+            reader.readLine();
```
어째서 첫줄을 무시하는지 의미부여하면 좋겠어요.

jinu0328 replied:
[refactor: csv 파일의 첫줄을 스킵하는 로직 메서드로 분리하여 의미 부여](https://github.com/woowacourse/java-attendance/pull/49/commits/04042dd1589845eb14915bb28df824b2b3f63756)
skipHeader라는 메서드를 분리하여 의미를 부여했습니다!

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+    private void loadAttendanceHistory(BufferedReader reader, Map<String, Crew> crewMap, List<Crew> crews) throws IOException {
+        String line;
+        while ((line = reader.readLine()) != null) {
+            String[] values = line.split(",");
+            String[] datetimeValues = values[1].split(" ");
+            String date = datetimeValues[0];
+            String time = datetimeValues[1];
+
+            Crew crew = getCrew(crewMap, crews, values[0]);
```
메서드를 분리하면 의미 전달이 뚜렷해질 듯 해요

jinu0328 replied:
[refactor: 문자열 파싱 및 도메인 객체 추가 메서드 분리](https://github.com/woowacourse/java-attendance/pull/49/commits/6372fa133e9fc8cf5e121603d38a18a073db0140)
분리 완료하였습니다!

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    private void processAttendanceTasks(String option) {
+        processAttendance(option);
+        processAttendanceEdit(option);
+        processAttendanceHistory(option);
+        processPenaltyCheck(option);
```
각 메서드의 구현을 들여다보니, 결국 if 문의 나열과 다르지 않네요. 오히려 if-else보다 성능적으로 좋지 않아요.
else 예약어를 사용하지 말라는 제약이 왜 주어졌을까요? 어떻게 하면 if-else 없이 분기로직을 처리할 수 있을지 고민해보고 적용해보세요. 객체지향의 다형성을 적극 활용해보시기 바랍니다~

jinu0328 replied:
[refactor: if-else 분기 로직을 Map과 Enum을 활용한 구조로 개선](https://github.com/woowacourse/java-attendance/pull/49/commits/b610786619646f1dc6e5bb4d1b84be9cc3c042f0)
Option이라는 Enum을 구현하고 컨트롤러에서 Map<Option, Runnable>을 통해 사용자 입력 분기를 개선하였습니다!

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,111 @@
+        if (!option.equals("2")) return;
+
+        Crew crew = crews.findByNickname(inputView.getEditNickname());
+        Attendance attendance = crew.findByDate(Converter.convertStringToInteger(inputView.getEditDayOfMonth()));
```
inputView으로부터 String 값을 전달 받아야만 하나요? int 값을 전달 받으면 안되나요? 입력에 대한 검증은 어디서 책임지는 것이 옳을까요?

jinu0328 replied:
지금까지는 View는 단순히 입·출력을 담당하는 역할이라고 생각했기 때문에, 모든 입력을 String으로 받고 유효성 검사나 문자열 파싱의 역할을 부여하지 않았었습니다.

그런데 이프의 리뷰를 보고 다시 생각해보니, 실제 프로젝트에서도 클라이언트에서 입력된 값을 서버가 요구하는 형식에 맞춰 변환하여 전달하는 것이 일반적이라는 점을 떠올리게 되었습니다.

이렇게 보면, View에서 미리 유효성 검사와 파싱을 거치는 것이 꼭 View의 역할을 벗어나는 것은 아닐 수도 있겠다는 생각이 들었습니다.

혹시 제가 내린 이 결론이 이프가 생각해보도록 유도한 방향과 일치하는지 여쭤보고 싶습니다! 😊

sinb57 replied:
저도 동일하게 생각해요. 결국 어딘가에 속해야 하는 책임인데, 지금의 Controller가 너무 무겁지는 않나 생각해보세요. Controller는 무슨 역할을 수행하는 친구로 구상하셨나요?

jinu0328 replied:
제가 생각하는 Controller의 역할은 다음과 같습니다.
- 사용자의 요청의 진입점
- 사용자의 요청에 따라 어떤 처리를 할지 결정
- 도메인의 비지니스 로직과 뷰의 입출력 사이의 다리 역할

위의 역할을 수행하는 것으로 충분하다고 했을때 이프의 피드백대로 Controller에서 입력에 대한 유효성 검사나 파싱을 거치는 것보다 View에서 담당하는 것이 맞다고 생각하여 리팩토링 하였습니다!
[refactor: 입력값 변환 로직을 InputView에서 처리하도록 변경](https://github.com/woowacourse/java-attendance/pull/49/commits/9a3b8e9f70e27d07f6bfc0bb6ba9496e7cbe48bc)

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,111 @@
+    public void processPenaltyCheck(String option) {
+        if (!option.equals("4")) return;
+
+        List<CrewDto> crewDtos = crews.createCrewDtos();
+        List<CrewDto> penaltyCrewDtos = crewDtos.stream()
+                .filter(crewDto -> crewDto.getPenaltyStatus() != PenaltyStatus.NONE)
+                .collect(Collectors.toCollection(ArrayList::new));
+
+        outputView.printPenaltyCrews(new CrewDtos(penaltyCrewDtos));
```
CrewDtos 가 필요한가요? penaltyCrewDtos 를 outputView에 전달하면 안되나요??

jinu0328 replied:
`CrewDtos`에 미션 조건에 따른 정렬 기능이 정의되어 있기 때문에, 해당 클래스를 사용하면서 단계가 하나 늘어났습니다.

이 부분에 대해서는 페어와도 논의했었는데, 결국 "정렬이라는 역할을 `View`에서 처리할지, `도메인`에서 처리할지" 선택해야 했습니다.
현재 정렬 조건이 단순한 문자열 오름차순, 정수 오름차순이 아니라 도메인 객체의 값과 연결되어 있어,
정렬 로직을 도메인에서 처리하는 것이 적절하다고 판단하여 `CrewDtos`에서 이 역할을 수행하도록 결정했습니다.

다만, 이로 인해 컨트롤러에서 `CrewDtos`를 한 번 더 생성해야 하면서 로직이 직관적이지 않고 무거워진 점은 사실입니다.
이런 경우, 도메인 객체의 값과 연관이 있는 정렬이라도 `View`에서 처리하는 것이 더 좋은 방식일까요?

sinb57 replied:
정렬은 무엇을 위한 조건인가요? 어떠한 변경이 있을 때 정렬이라는 행위에 변화가 생길까요? 그러한 변경에 도메인이 변경되어야 할까요, view가 변경되어야 할까요? 비즈니스 로직과 UI 로직을 구분 지은 이유는 무엇일까요?

jinu0328 replied:
정렬은 결국 사용자가 데이터를 볼 때의 편의성을 위한 조건이며, 출력 형식이 달라질 때 정렬의 방식도 변할 가능성이 크다고 생각합니다.

처음에는 정렬이 데이터를 다루는 과정에서 발생하므로 비즈니스 로직에 해당하는 것이 아닐까 고민했지만,
정렬이 필요한 이유를 다시 생각해보니, 출력의 가독성을 높이기 위한 역할이 더 크다는 점을 깨달았습니다.

비지니즈 로직과 UI 로직을 구분한 이유는 결국 핵심 비지니스 로직을 지켜주기 위해서라고 생각합니다. UI의 요구사항에 따라 비지니스 로직이 달라진다면 안전성이 높은 코드라고 볼 수 없을 것 같습니다.

이프의 피드백 덕분에 앞으로도 정렬과 같은 기능을 마주할 때, 이것은 도메인의 핵심 규칙인가, UI의 요구사항인가?를 더 잘 따져볼 수 있을 것 같습니다. 감사합니다!

[refactor: 제적대상자 CrewDto를 정렬하기 위해 구현했던 CrewDtos 제거](https://github.com/woowacourse/java-attendance/pull/49/commits/9f69b4cf96ad6643b55ab4bceb9973691a8deb77)

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,56 @@
+            isAbsent = true;
+            return;
+        }
+        isLate = isLate();
+        isAbsent = isAbsent();
+    }
+
+    private boolean isLate() {
+        return day.isLate(attendanceTime);
+    }
+
+    private boolean isAbsent() {
+        return day.isAbsent(attendanceTime);
+    }
+
```
메서드를 분리하여 얻은 장점이 궁금합니다.

jinu0328 replied:
관성적으로 분리 가능한 부분이라고 생각하여 메서드 분리를 진행했던 것 같습니다. 다시 살펴보니 그대로 사용하여도 충분히 직관적인 로직이고 별도로 재사용되지 않으니 하나의 메서드에서 직접 처리하는 것이 적절할 것 같습니다!

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    private Integer calculateNonAttendanceCount() {
+        return calculateAbsentCount() + calculateLateCount() / 3;
```
매직넘버가 보이네요

jinu0328 replied:
Attendances를 새롭게 구현하며 상수로 정의하였습니다!
[refactor: Attendance의 일급 컬렉션 Attendances 구현](https://github.com/woowacourse/java-attendance/pull/49/commits/dd0bb95e8c300d0651ceddf301aaaa0ee1cdf3ab)

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,98 @@
+
+    private Integer calculateAbsentCount() {
+        return (int) attendances.stream()
+                .filter(attendance -> attendance.toDto().getAbsent().equals(true))
```
왜 Dto로 변환하여 getter를 통해 값을 가져다 비교하는지 궁금합니다.
attendance.isAbsent()하면 되지 않을까요?
제프가 생각하는 DTO의 역할은 무엇인가요??

jinu0328 replied:
제가 생각하는 현재 애플리케이션에서의 `DTO`의 역할은 도메인 객체의 값을 안전하게 `View`로 전달하기 위해, 즉 `View`가 도메인 객체를 직접 참조하지 못하도록 보호하는 "얇은 방파제" 역할을 수행하는 것이라고 생각합니다.

다만, 피드백 주신 부분을 다시 살펴보니 현재 `getter`를 통해 비교하는 로직은 뷰가 아닌 도메인 객체 간의 상호작용에서 발생하는 것이므로,
굳이 `DTO`로 변환하여 값을 가져오는 대신, 도메인 객체의 `getter`를 직접 활용하는 것이 더 적절할 것 같다는 생각이 들었습니다!

sinb57 replied:
답변 주신 DTO에 대한 내용에 대해 확인하고 싶은 부분이 있어서 추가 질문 드려보아요~
에드가 생각하는 DTO는 "계층 간 데이터 전달을 위한 객체"가 맞을까요? DTO는 "계층 간"에 국한된 객체일까요?

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,98 @@
+        return this.nickName.equals(nickname);
+    }
+
+    public Boolean isAlreadyAttend(LocalDate date) {
+        for (Attendance attendance : attendances) {
+            if (attendance.isEqualTo(date)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public void recordAbsence() {
+        LocalDate today = LocalDate.now();
+
+        for (LocalDate date = today.withDayOfMonth(1); date.isBefore(today); date = date.plusDays(1)) {
+            if (date.getDayOfWeek().getValue() == 6 || date.getDayOfWeek().getValue() == 7) {
+                continue;
+            }
+            if (!isAlreadyAttend(date)) {
+                addAttendance(new Attendance(new Day(date), null));
+            }
+        }
+    }
```
> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.

요구사항을 지켜주세요~

sinb57 commented on src/main/java/domain/CrewDto.java:
```diff
@@ -0,0 +1,32 @@
+package domain;
+
+public class CrewDto {
```
DTO를 왜 만들었는지 궁금해요.
DTO는 무엇이며, 지금 어떠한 목적으로 사용하고 있나요?

jinu0328 replied:
현재 애플리케이션에서의 DTO의 역할은 도메인 객체의 값을 안전하게 View로 전달하기 위해, 즉 View가 도메인 객체를 직접 참조하지 못하도록 보호하는 "얇은 방파제" 역할을 수행하고 있습니다! 추가로 직접적인 도메인 객체의 멤버 값이 아니어도 뷰에서 요구하는 값을 미리 계산하여 Dto에 보관하고 있습니다!

sinb57 replied:
에드가 생각하는 DTO는 "계층 간 데이터 전달을 위한 객체"가 맞을까요? DTO는 "계층 간"에 국한된 객체일까요? "계층"이란 무엇일까요?

jinu0328 replied:
DTO(Data Transfer Object)가 일반적으로 "계층 간 데이터 전달" 위해 쓰이는 것은 맞다고 생각합니다. 그러나 "계층 간"에 국한된다고 한다면 아니라는 생각이 듭니다. DTO는 단순히 "계층 간" 전달을 위한 객체가 아니라, 데이터의 구조를 명확히 하고, 특정 목적에 맞는 데이터를 전달하는 역할을 하는 객체라고 보는 것이 더 적절한 것 같습니다.
이번 미션에서의 DTO는 앞서 언급한 대로 View와 도메인 사이의 얇은 방파제 역할을 수행하며 View가 필요로 하는 데이터를 미리 가공하여 전달하는 역할을 맡고있습니다. 

아직 이프가 질문을 통해 전달하고자 하는 바를 제가 정확히 캐치하지 못한 것 같습니다. 😢  추가 피드백 가능할까요?

sinb57 replied:
DTO는 단순히 view와 domain 사이의 데이터 전달을 목적으로 하지 않다는 것을 알려드리고 싶었어요~ 답변 주셔서 감사해요ㅎ

그럼 이어서 추가 질문드릴게요~ 에드는 CrewDto를 입출력을 위한 Dto라고 하셨어요. 그런데 지금 보면 DTO의 생성 로직이 Crew 도메인에 존재해요. 도메인이 DTO에 의존해도 괜찮을까요?

sinb57 commented on src/main/java/util/Converter.java:
```diff
@@ -0,0 +1,39 @@
+
+    public static LocalTime convertStringToLocalTime(String timeInput) {
+        try {
+            DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm");
```
메서드가 호출될 때마다 객체가 생성되는데 재사용할 수 없을까요?

jinu0328 replied:
formatter를 상수로 지정하여 재사용하도록 했습니다!
[refactor: DateTimeFormatter 객체를 상수로 지정하여 util 내부 메서드에서 재사용](https://github.com/woowacourse/java-attendance/pull/49/commits/3ad0db55c0a51cbb3e833d2eee8ba92e5c38be98)

sinb57 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Scanner;
+
+public class InputView {
+    private static final String NICKNAME_INPUT_MESSAGE = "닉네임을 입력해 주세요.";
+    private static final String ATTENDANCE_INPUT_MESSAGE = "등교 시간을 입력해 주세요.";
+    private static final String EDIT_NICKNAME_INPUT_MESSAGE = "출석을 수정하려는 크루의 닉네임을 입력해 주세요.";
+    private static final String EDIT_DAY_OF_MONTH_INPUT_MESSAGE = "수정하려는 날짜(일)를 입력해 주세요.";
```
문자열을 재사용하지 않음에도 상수로 분리한 이유가 무엇인가요?

jinu0328 replied:
기본적으로 출력해야 하는 문자열은 상수로 처리하는 것이
추후 수정이 용이하고, 코드의 가독성 측면에서도 유리할 것이라고 판단했기 때문입니다.

다만, 다시 살펴보니 오히려 상수화한 문자열이 코드 가독성을 높인다고 보기는 어려울 수도 있을 것 같습니다.
특히, 출력 문자열이 짧거나 의미가 명확한 경우에는 그대로 사용하는 것이 더 직관적일 수도 있겠네요.

상수 처리를 할 때의 기준을 좀 더 명확하게 정리해보겠습니다!

sinb57 reviewed (changes requested):
안녕하세요 에드~ 출석 미션의 리뷰어를 맡은 이프입니다!
이번에 같이 매칭된 제프와 페어였나봐요!! 동일한 부분에 대해 동일 피드백 드려 보아요~
질문에 대한 답변과 코드에 대한 코멘트 확인 부탁드려요~

공통 피드백) 
열려있는 모든 메서드에 대해 최대한 단위 테스트를 작성해주시고,
아래의 프로그래밍 요구사항에 유의하여 수정해주세요.
- indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.
- 함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.
- else 예약어를 쓰지 않는다.

질문)
> 선택한 점수의 이유를 적어주세요.
> 도메인에서의 getter, setter를 일부 허용하였기 때문입니다..

setter를 사용하셨다고 했는데, 어디에 있나요??

---

남겨주신 질문에 답변 드리려고 보니, 제프의 질문과 동일하여.. 동일 답변 드려보아요.

> 1. 도메인 설계 및 테스트 전략 관련 질문

사람마다 스타일이 다르겠지만, 저는 후자의 방식을 선호해요. 요구사항에 기반하여 대략적인 설계를 먼저 구상하고서 접근하다보니, 어떠한 객체들이 필요하고 어떠한 책임과 역할을 부여하면 좋을지 판단이 되거든요. 이렇게 기초를 쌓은 상태에서 테스트코드와 함께 구현에 대한 즉각적인 피드백을 주고받으며 개발하곤 합니다.

에드가 생각하는 테스트코드의 장점과 역할은 무엇인가요?

> 2. List<Attendance>에 대한 설계 고민 질문

List<Attendance>에 행위를 부여하여 응집력을 높이고 역할을 부여할 수 있다면, 충분히 Attendances를 만들 이유가 될텐데요. 장단점을 보던 중 의아한 부분이 있어서 질문드립니다. 왜 일급컬렉션은 항상 불변을 보장해야 하나요??

> 3. Day 객체를 만든 이유 및 유지 여부에 대한 고민

공휴일을 담지 않는, 즉 "주중"의 의미를 지닌 객체로 역할을 부여하면 어떨까 싶어요. 공휴일로 객체를 생성할 수 없도록 하면 되지 않을까요? 실제로 현재 `Day#checkHoliday` 메서드가 공휴일 검증에만 사용되는 것을 볼 수 있고, 요구사항도 공휴일에 대한 제약이 있다보니 Day 객체의 역할을 구체적으로 바꾸어봐도 좋아보여요.

jinu0328 commented:
> setter를 사용하셨다고 했는데, 어디에 있나요??

Attendance 클래스의 다음 메서드입니다!
```java
    public void updateAttendanceTime(LocalTime attendanceTime) {
        this.attendanceTime = attendanceTime;
        updateStatus();
    }
```


> > 1. 도메인 설계 및 테스트 전략 관련 질문
> 
> 사람마다 스타일이 다르겠지만, 저는 후자의 방식을 선호해요. 요구사항에 기반하여 대략적인 설계를 먼저 구상하고서 접근하다보니, 어떠한 객체들이 필요하고 어떠한 책임과 역할을 부여하면 좋을지 판단이 되거든요. 이렇게 기초를 쌓은 상태에서 테스트코드와 함께 구현에 대한 즉각적인 피드백을 주고받으며 개발하곤 합니다.
> 
> 에드가 생각하는 테스트코드의 장점과 역할은 무엇인가요?

제가 이번 미션을 수행하며 체감한 테스트 코드의 장점과 역할은 다음과 같습니다.

- 작은 단위로 테스트를 정의하면, 프로덕션 코드에서는 해당 테스트를 통과하는 데 집중할 수 있었습니다. 처음부터 너무 넓게 보지 않고 필요한 코드만 작성할 수 있어 불필요한 로직이 줄어드는 효과가 있었던 것 같습니다.

- 새로운 기능을 추가할 때 기존 테스트가 실패하면, 이전 코드와 충돌이 발생했다는 신호를 즉각적으로 받을 수 있었습니다. 이를 통해 잘못된 설계를 빠르게 캐치하고 수정할 수 있었습니다.

- 반대로 기존 테스트가 정상적으로 동작한다면, 리팩토링 후에도 기존 기능이 유지된다는 확신을 가질 수 있어 코드 개선을 두려움 없이 시도할 수 있는 환경이 만들어졌다고 생각합니다.

- 추가적으로 테스트 코드는 단순히 버그를 잡는 도구가 아니라, 해당 기능이 어떻게 동작해야 하는지를 보여주는 명세서 역할도 한다고 생각합니다. 

다만 테스트 코드를 처음 작성하다 보니 어느 단위에서부터 시작해야 할지 감이 잡히지 않아 생산성이 떨어지는 느낌을 받았습니다. 또한, 어느 정도 도메인을 설계한 상태에서 코드를 작성하다 보니, 작은 단위로 테스트를 작성하려 해도 본능적으로 앞으로 작성할 로직까지 신경을 쓰게 되어 쉽지 않았던 것 같습니다. 별도의 도메인 설계 없이 테스트 코드를 작성하며 도메인 설계에 들어가는 것이 맞을까요?

> > 2. List에 대한 설계 고민 질문
> 
> List에 행위를 부여하여 응집력을 높이고 역할을 부여할 수 있다면, 충분히 Attendances를 만들 이유가 될텐데요. 장단점을 보던 중 의아한 부분이 있어서 질문드립니다. 왜 일급컬렉션은 항상 불변을 보장해야 하나요??

말씀해주신 부분을 다시 생각해보니, 일급 컬렉션이 반드시 불변성을 보장해야 하는 것은 아니라는 점을 간과했던 것 같습니다. 제가 일급 컬렉션을 적용하지 않은 이유는 다음과 같습니다. 

1. 일급 컬렉션에 대한 일반적인 기대와의 불일치
    - 일반적으로 일급 컬렉션을 보면, 팀원들은 "이 객체는 불변일 것이다"라고 예상할 가능성이 높습니다.
    - 하지만 Attendances가 계속해서 출석 데이터를 추가/수정해야 하는 객체라면, 불변성을 유지하기 어렵고 결국 `add()`, `update()` 같은 변경 메서드가 필요하게 됩니다.
    - 이 경우, 팀원들이 "Attendances는 불변일 것이다"라는 잘못된 기대를 가지면, 사용 시점에서 혼란이 생길 수 있고, 코드의 의도를 정확히 파악하기 어려워질 수 있습니다.

2. 불필요한 리소스 소모 방지
    - 매번 컬렉션을 일급 컬렉션으로 변환하는 과정 자체가 불필요한 오버헤드가 될 수도 있다고 생각했습니다.

3. 차라리 컬렉션이 변경될 수 있음을 명확히 드러내는 것이 낫다고 판단
    - 불변성을 유지할 수 없는 객체라면, 애초에 불변성을 기대하지 않는 `List<Attendance>`를 사용하는 것이 더 직관적일 수 있다고 생각했습니다.
    - 예를 들어, 팀원이 Attendances를 보고 "이건 불변할 거야"라고 생각했는데, 내부적으로 `addAttendance()` 같은 메서드가 존재한다면, 사용자가 혼란을 겪을 가능성이 큽니다.
    - 반면, `List<Attendance>`를 그대로 사용하면 "이 컬렉션은 변경 가능하다"는 점이 더 명확하게 드러나므로, 불필요한 오해를 줄일 수 있다고 생각했습니다.

> > 3. Day 객체를 만든 이유 및 유지 여부에 대한 고민
> 
> 공휴일을 담지 않는, 즉 "주중"의 의미를 지닌 객체로 역할을 부여하면 어떨까 싶어요. 공휴일로 객체를 생성할 수 없도록 하면 되지 않을까요? 실제로 현재 `Day#checkHoliday` 메서드가 공휴일 검증에만 사용되는 것을 볼 수 있고, 요구사항도 공휴일에 대한 제약이 있다보니 Day 객체의 역할을 구체적으로 바꾸어봐도 좋아보여요.

공휴일, 주말의 경우 Day 객체가 아예 생성되지 않게 하는 방법이 있었군요! 이프의 피드백대로 수정해보겠습니다!!

jinu0328 committed:
chore: AttendanceHistoryLoader 클래스의 패키지 config -> loader로 변경

jinu0328 committed:
refactor: csv 파일의 첫줄을 스킵하는 로직 메서드로 분리하여 의미 부여

jinu0328 committed:
refactor: 문자열 파싱 및 도메인 객체 추가 메서드 분리

jinu0328 committed:
refactor: if-else 분기 로직을 Map과 Enum을 활용한 구조로 개선

- Option enum 추가 구현
- 기존 if-else로 처리되던 사용자 입력 분기를 Map<Option, Runnable>을 활용하여 개선

jinu0328 committed:
refactor: 입력값 변환 로직을 InputView에서 처리하도록 변경

- 기존 컨트롤러에서 String 입력을 받아 변환하는 방식에서 InputView에서 변환된 값을 반환하도록 개선
- 컨트롤러에서의 불필요한 변환 로직 제거하여 코드 가독성 및 책임 분리 강화

jinu0328 committed:
refactor: 제적대상자 CrewDto를 정렬하기 위해 구현했던 CrewDtos 제거

- 기존 CrewDtos를 생성하여 정렬했던 방식을 View에서 직접 전달 받은 List<CrewDto>를 정렬하는 것으로 변경

jinu0328 committed:
refactor: 불필요한 메서드 분리 제거

jinu0328 committed:
refactor: Crew의 private 메서드에서 사용되는 매직넘버 제거

jinu0328 committed:
feat: 도메인끼리의 값 전달에 있어 dto를 통하지 않고 조회할 수 있도록 수정

jinu0328 committed:
refactor: stream을 활용하여 기존 메서드의 depth 1로 조정

jinu0328 committed:
refactor: DateTimeFormatter 객체를 상수로 지정하여 util 내부 메서드에서 재사용

jinu0328 committed:
refactor: View의 안내메세지 상수 제거

- 재사용되지 않고 크게 가독성을 향상 시키지 않는다고 판단되어 상수처리하지 않고 메서드에서 직접 정의

jinu0328 committed:
refactor: 공휴일의 날짜를 저장하는 Holiday enum 구현

- Localdate 객체를 파라미터로 받아 공휴일 목록에 존재하는 날짜인지 확인하는 isHoliday 메서드 구현

jinu0328 committed:
refactor: 생성자에 checkOffDay 메서드를 호출하여 휴일 객체 생성 방지

jinu0328 committed:
refactor: Day 객체 생성 시 휴일에 대한 검사가 이루어지므로 Attendance의 휴일 검증 로직 제거

sinb57 reviewed (changes requested):
안녕하세요 에드~ 작업하신 내용은 반영 안된 것 같아요!!
커밋하신 내용들을 push 해주시면 좋겠어요~ㅎ
그래도 남겨주신 답변과 질문 확인했고, 코멘트 남겼으니 확인 부탁드려요 ~

---

> 2024년 12월 한 달 동안 시범적으로 최소한의 기능을 갖춘 출석 시스템을 개발하여 출석을 체계적으로 관리하기로 하였습니다

요구사항을 보면, 현재는 12월 한정 시범 운영이지만 시스템이 확장될 가능성을 엿볼 수 있어요. 이 경우 현재의 구현에서 확장에 대응하려면 어떠한 변경이 있을지 생각해보시겠어요? 과연 현재의 코드는 변경에 열려 있나요?

---

> 도메인 설계 및 테스트 전략 관련 질문

`작은 단위로 테스트를 정의하면, 프로덕션 코드에서는 해당 테스트를 통과하는 데 집중할 수 있었습니다. 처음부터 너무 넓게 보지 않고 필요한 코드만 작성할 수 있어 불필요한 로직이 줄어드는 효과가 있었던 것 같습니다.`

테스트 코드의 피드백에는 여럿 관점이 있지만, 그 중 구현에 대한 피드백이 존재한다고 봐요. 답변 주신 위의 내용도 테스트 코드가 지닌 구현 관점의 장점이죠. 또 다른 장점은 없을까요? 에드의 고민에 큰 도움이 될 거예요!

TDD의 첫 스텝으로 실패하는 테스트 코드를 작성하잖아요. 어째서 TDD는 실패하는 테스트 코드를 먼저 작성할까요?

jinu0328 committed:
refactor: AttendanceHistoryLoader의 패키지 loader -> domain으로 변경

jinu0328 committed:
refactor: Attendance의 일급 컬렉션 Attendances 구현

- Crew에서 책임지던 Attendance 관련 메서드를 Attendances에서 수행하도록 변경

jinu0328 committed:
refactor: view에서 crew를 직접 참조하지 않고 crewDto를 참조하도록 변경

jinu0328 committed:
chore: 사용하지 않는 상수 제거

jinu0328 committed:
refactor: Crew에서 수행하던 findByDate의 유효성 검사 메서드 Attendances에서 수행하도록 변경

jinu0328 committed:
refactor: 메서드 depth 1로 수정

jinu0328 committed:
refactor: Crews에 addCrew 메서드를 추가하여 Loader의 로직 간편화

jinu0328 committed:
refactor: AttendanceHistoryLoaderTest의 테스트 구현

- 기존 메서드 내에서 정의됐던 FileReader를 파라미터로 받도록 수정
- FileReader를 파라미터로 받음으로써 여러 csv 파일의 형식을 테스트할 수 있도록 함

jinu0328 commented:
> 요구사항을 보면, 현재는 12월 한정 시범 운영이지만 시스템이 확장될 가능성을 엿볼 수 있어요. 이 경우 현재의 구현에서 확장에 대응하려면 어떠한 변경이 있을지 생각해보시겠어요? 과연 현재의 코드는 변경에 열려 있나요?

현재 구현된 코드에서 시스템의 확장을 고려한다면 다음과 같은 변경을 수행하면 대응할 수 있다고 생각합니다.
- Holiday enum에 지정된 공휴일을 추가
- 사용자에게 날짜를 입력 받을 시 연(year), 월(month)도 추가로 받기
- findByDate 메서드에서 dayOfMonth뿐만 아니라 연(year), 월(month)도 고려하여 Attendance 조회하기

이렇게 보니 결국 입력 형식에 따라 도메인의 비지니스 로직이 바뀌게 되는 결과가 나온 것 같습니다..!
확장성을 고려하여 findByDate와 같이 현재 일(dayOfMonth)만을 고려하는 기능을 리팩토링 해두는 것이 좋을까요?

> 테스트 코드의 피드백에는 여럿 관점이 있지만, 그 중 구현에 대한 피드백이 존재한다고 봐요. 답변 주신 위의 내용도 테스트 코드가 지닌 구현 관점의 장점이죠. 또 다른 장점은 없을까요? 에드의 고민에 큰 도움이 될 거예요!
> TDD의 첫 스텝으로 실패하는 테스트 코드를 작성하잖아요. 어째서 TDD는 실패하는 테스트 코드를 먼저 작성할까요?

먼저 실패하는 테스트를 작성하는 과정에서 "이 기능이 어떻게 동작해야 하는지" 고민하며 요구사항을 정리하게 되는 효과가 있다고 생각합니다.
앞서 언급했던 구현 관점의 장점이 아닌 다른 장점을 생각해본다면 **리팩토링에 대한 두려움을 줄여준다** 라는 장점도 존재할 것 같습니다.
현재는 미션을 수행하는 것이기에 잘못된 코드를 짠다고 해서 리스크가 존재하지 않지만 현업에서 정말 중요한 비지니스 로직을 추가, 수정할 때는 큰 부담이 따를 수 있습니다. 만약 TDD 사이클을 따르지 않는다면 수정된 코드가 기존 기능을 깨뜨릴 위험이 있으며 이를 사전에 방지하기 어렵습니다. 테스트 코드를 먼저 작성함으로써 기능 요구사항을 정리하고 이를 리팩토링을 통해 통과하도록 수정한다면 결과물이 더욱 안정적이고 개발자 자신도 큰 부담없이 리팩토링을 진행 할 수 있을 것 같습니다.

sinb57 commented on src/test/java/domain/AttendanceHistoryLoaderTest.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Test
+    void 유효한_형식의_파일에_대해_Crew를_정상적으로_초기화한다() throws Exception {
+        String csvData = "nickname,datetime\n에드,2025-02-13 08:30\n제프,2025-02-14 09:15\n";
```
```suggestion
        String csvData = """
                nickname,datetime
                에드,2025-02-13 08:30
                제프,2025-02-14 09:15
                """;
```
텍스트 블럭을 활용하면 좋아요~
https://www.baeldung.com/java-text-blocks

jinu0328 replied:
감사합니다! 적용 완료 하였습니다

sinb57 commented on src/test/java/domain/AttendanceHistoryLoaderTest.java:
```diff
@@ -0,0 +1,50 @@
+
+public class AttendanceHistoryLoaderTest {
+
+    private File createTempCsvFile(String csvData) throws IOException {
+        Path tempFile = Files.createTempFile("test_attendance", ".csv");
+        Files.write(tempFile, csvData.getBytes());
+
+        return tempFile.toFile();
+    }
```
👍

sinb57 commented on src/main/java/domain/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,63 @@
+            skipHeader(reader);
+            loadAttendanceHistory(reader, crewMap, crews);
+        } catch (IOException e) {
+            e.printStackTrace();
```
`printStackTrace`는 사용을 지양하는게 좋아요

jinu0328 replied:
printStackTrace는 예외 추적에 오버헤드가 발생하고 파악이 어렵다는 단점이 있다는 걸 알게되었습니다!
지정된 메세지로 예외를 발생시키도록 수정하였습니다.
[refactor: IOException 메시지 추가](https://github.com/woowacourse/java-attendance/pull/49/commits/9bc0de1d9177f5b73a54e2d7b76b61b86a28be8e)

sinb57 commented on src/main/java/domain/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,63 @@
+        String date = datetimeValues[0];
+        String time = datetimeValues[1];
+
+        Crew crew = getCrew(crewMap, crews, values[0]);
```
변수 분리를 통해 values[0]에 의미 부여하면 어떨까요?

jinu0328 replied:
완료하였습니다!
[refactor: 변수 분리를 통해 values[0]에 의미 부여](https://github.com/woowacourse/java-attendance/pull/49/commits/e095ad9e35a9c31ca113d37eb78a4db226b83f03)

sinb57 commented on src/main/java/domain/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,63 @@
+    private final DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HH:mm");
+
+    public Crews loadCrews(FileReader fileReader) {
+        Crews crews = new Crews();
+        Map<String, Crew> crewMap = new HashMap<>();
```
로직을 살펴보니, crewMap와 crews가 모든 private 메서드로 깊숙이 파고 들어가네요. Map 자료구조가 필요한가요? Crews 객체를 활용해서 책임을 위임하면 어떨까요?

jinu0328 replied:
Crews에 add 메서드를 추가한 뒤로는 Map을 이용할 필요가 없었는데 리팩토링 해야하는 부분을 놓쳤던 것 같습니다..!
Crews 객체만으로 처리하도록 수정하였습니다.
[refactor: crewMap 제거 및 crews에 Crew 객체 반환 책임 위임](https://github.com/woowacourse/java-attendance/pull/49/commits/1af190a373756c217e45294111b02e368ca71fbb)

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,120 @@
+    private final Crews crews;
+    private final InputView inputView;
+    private final OutputView outputView;
+    private final Map<Option, Runnable> options = Map.of(
+            Option.ATTEND, this::processAttendance,
+            Option.EDIT_ATTENDANCE, this::processAttendanceEdit,
+            Option.SHOW_ATTENDANCE_HISTORY, this::processAttendanceHistory,
+            Option.SHOW_PENALTY_CREWS, this::processPenaltyCheck,
+            Option.EXIT, this::exitApplication
+    );
```
Option Enum이랑 합쳐도 좋지않을까요?

jinu0328 replied:
Consumer를 이용하여 Option enum이 Controller의 메서드 실행을 담당하도록 리팩토링하였습니다.
[refactor: Option enum이 Consumer를 활용해 AttendanceController의 메서드 실행을 담당하도록 변경](https://github.com/woowacourse/java-attendance/pull/49/commits/3e3d45edf28f6f0cf9c3a7eae10199a3b66b1c1d)

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    private void exitApplication() {
+        System.exit(0);
```
`System.exit(0);`은 사용하지 말아주세요. 몹시 위험한 방식입니다.

jinu0328 replied:
System.exit(0)은 JVM 자체를 강제 종료 시키며 여러 상정치 못한 오류를 불러올 수 있다는 점 알게되었습니다..!
수정 완료하였습니다

sinb57 commented on src/main/java/util/Converter.java:
```diff
@@ -0,0 +1,37 @@
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+
+public class Converter {
```
Converter가 이제는 view에서만 사용하네요! util로 존재해야 할까요? view 안에 private으로 존재해도 되지 않을까요?

jinu0328 replied:
기존 Converter에 util 클래스임에도 
```java
public static String covertLocalTimeToString(LocalTime localTime) {

        if (localTime == null) {
            return "--:--";
        }

        return localTime.format(TimeFormat.DATE_TIME_FORMATTER);
    }
```
위와 같이 뷰의 특정 형식에 의존하는 형태인 것이 어색하게 느껴졌었습니다.
이프의 말대로 view에서만 사용하는 만큼 view 안의 private 메서드로 구현하였습니다! 
다만 DATE_TIME_FORMATTER(HH:mm)의 경우 InputView와 OutputView에서 재사용할 수 있도록 TimeFormat이라는 별도의 클래스에서 상수로 관리하도록 하였습니다.
[refactor: Converter 클래스 제거 및 기존 기능 view의 내부 메서드로 구현](https://github.com/woowacourse/java-attendance/pull/49/commits/095570bc6b3f33ff67b9e854910ace5d333a43b7)

sinb57 commented on src/main/java/domain/Attendances.java:
```diff
@@ -0,0 +1,79 @@
+    public Attendances() {
+    }
+
+    public void addAttendance(Attendance attendance) {
```
메서드 시그니처를 봤을 때, 메서드명의 `Attendance`와 `파라미터명의 `Attendance가 중복으로 존재하네요. `public void add(Attendance attendance)` 이어도 충분해 보이는데 어떻게 생각하시나요? 외부의 시점에서는 `attendances.add(attendance)`로 사용하겠네요.

jinu0328 replied:
동의합니다! 메서드명과 파라미터명의 중복을 제거함으로써 가독성이 더 좋아진 것 같습니다.
리팩토링 완료하였습니다.

sinb57 reviewed (changes requested):
안녕하세요 에드~ 피드백 반영해주셔서 감사해요~~ 추가 코멘트 남겼으니 확인 부탁드려요ㅎ

---

> 현재 구현된 코드에서 시스템의 확장을 고려한다면 다음과 같은 변경을 수행하면 대응할 수 있다고 생각합니다.
> ...
> 이렇게 보니 결국 입력 형식에 따라 도메인의 비지니스 로직이 바뀌게 되는 결과가 나온 것 같습니다..!확장성을 고려하여 findByDate와 같이 현재 일(dayOfMonth)만을 고려하는 기능을 리팩토링 해두는 것이 좋을까요?

잘 파악하셨어요!! 때론 변경 가능성을 미리 수용해서 열린 설계를 해도 좋아요. 하지만 때론 오버 엔지니어링이 되어 유지보수에 문제를 일으키기도 합니다. 그렇기 때문에 당장 할 수 있는 최소한의 대응은, 변화에 대응하기 위해 어떠한 변경이 필요한지 사이드이펙트를 미리 예측하고 있는 것이에요. 언제 변화가 발생할지 모르기 때문이에요. 

> 먼저 실패하는 테스트를 작성하는 과정에서 "이 기능이 어떻게 동작해야 하는지" 고민하며 요구사항을 정리하게 되는 효과가 있다고 생각합니다. 앞서 언급했던 구현 관점의 장점이 아닌 다른 장점을 생각해본다면 리팩토링에 대한 두려움을 줄여준다 라는 장점도 존재할 것 같습니다. 현재는 미션을 수행하는 것이기에 잘못된 코드를 짠다고 해서 리스크가 존재하지 않지만 현업에서 정말 중요한 비지니스 로직을 추가, 수정할 때는 큰 부담이 따를 수 있습니다. 만약 TDD 사이클을 따르지 않는다면 수정된 코드가 기존 기능을 깨뜨릴 위험이 있으며 이를 사전에 방지하기 어렵습니다. 테스트 코드를 먼저 작성함으로써 기능 요구사항을 정리하고 이를 리팩토링을 통해 통과하도록 수정한다면 결과물이 더욱 안정적이고 개발자 자신도 큰 부담없이 리팩토링을 진행 할 수 있을 것 같습니다.

답변 주셔서 감사해요~ 말씀 주신 내용을 보면, "요구사항 파악"과 "안정적인 리팩터링"으로 요약할 수 있네요ㅎ

우선 "안정적인 리팩터링"의 경우, 실패하는 테스트가 아닌 성공하는 테스트가 필요하기 때문에 TDD가 아니어도 된다고 봐요. 리팩터링 후의 코드에 문제 없는지 확인하기 위해서는 기존 코드에서의 성공하는 테스트가 필요하기 때문이에요.

그리고 "요구사항 파악"은 음 맞아요ㅎ 이 또한 구현 관점의 장점이네요~ 제가 생각한 구현 관점의 장점은 "외부 시점으로 바라본다"예요. ~구체적으로 설명하고 싶지만, 오히려 여러분의 학습 방향에 영향을 줄까봐 말은 아낄게요.~ 혹시 궁금하시다면 2단계 진행 후 언제든 편하게 DM 주세요.

jinu0328 committed:
refactor: IOException 메시지 추가

jinu0328 committed:
refactor: 변수 분리를 통해 values[0]에 의미 부여

jinu0328 committed:
refactor: crewMap 제거 및 crews에 Crew 객체 반환 책임 위임

jinu0328 committed:
chore: 임시로 지정한 공휴일 날짜 수정

jinu0328 committed:
refactor: AttendanceDto 제거 및 방어적 복사를 이용한 getter 추가

- Attendances, Attendance의 복사 생성자 구현

jinu0328 committed:
refactor: CrewDto 제거 및 방어적 복사를 이용한 getter 추가

- Crew, Crews의 복사 생성자 구현

jinu0328 committed:
refactor: 뷰 및 테스트에서 도메인 객체를 참조하도록 변경

- 기존 도메인을 dto로 변환하고 이를 참조하게 한 방식에서 도메인 객체를 직접 전달

jinu0328 committed:
refactor: Option enum이 Consumer를 활용해 AttendanceController의 메서드 실행을 담당하도록 변경

- System.exit(0) 사용 철회

jinu0328 committed:
refactor: Converter 클래스 제거 및 기존 기능 view의 내부 메서드로 구현

- DateTimeFormatter를 상수로 관리하는 TimeFormatter 구현

jinu0328 committed:
chore: 텍스트 블럭을 통해 가독성 향상

jinu0328 commented:
## https://github.com/woowacourse/java-attendance/pull/49#discussion_r1967991613 에 대한 답변
처음에는 DTO가 View의 출력 형식에 의존하고 있다는 점을 문제 삼지 않았습니다.
DTO의 역할이 애초에 출력을 위한 데이터 전달이었기 때문입니다.
하지만 이프의 피드백을 통해, 현재 도메인이 DTO에 의존하고 있음을 깨닫게 되었습니다.
즉, DTO가 변경되면 도메인의 toDto() 메서드도 함께 변경될 가능성이 존재하며,
결과적으로 View의 변경이 도메인에 영향을 미치는 구조가 형성되고 있었습니다.

### 시도해본 것
- DTO에서 from() 메서드를 활용하여 변환 책임을 DTO 내부로 이동
- Mapper 클래스를 도입하여 DTO와 도메인 간의 변환을 분리
- 출력에 필요한 데이터를 도메인이 아닌 View에서 직접 가공하도록 변경

이 과정에서 DTO의 존재 이유에 대해 다시 고민하게 되었습니다.

### 처음 DTO를 도입한 이유
- 도메인 객체를 직접 노출하지 않고, 필요한 데이터만 전달하기 위한 용도
- View에서 요구하는 특정 형태의 데이터를 가공하여 전달하기 위한 용도

위의 이유를 고려했을 때, View에서 직접 출력 형태에 따라 데이터를 가공하는 방식으로 바꾼 이후, 현재 미션에서는 View가 도메인의 모든 멤버를 필요로 하고 있었으며, DTO가 하는 일이 단순한 값 복사에 불과했습니다.

따라서 DTO를 유지할 필요 없이, 도메인 객체의 방어적 복사만으로 충분하다고 판단하였습니다.

그 결과 Dto 클래스는 모두 제거하였고 domain 객체의 getter는 방어적 복사를 수행하고 있으며 이를 View에서 직접 참조하는 것으로 리팩토링 하였습니다
[refactor: AttendanceDto 제거 및 방어적 복사를 이용한 getter 추가](https://github.com/woowacourse/java-attendance/pull/49/commits/439350948c89131fc750534e13c4e0e4d2b461a3)
[refactor: CrewDto 제거 및 방어적 복사를 이용한 getter 추가](https://github.com/woowacourse/java-attendance/pull/49/commits/d0dbd479be98400af25a63a31c348c6c8d4abe5e)
[refactor: 뷰 및 테스트에서 도메인 객체를 참조하도록 변경](https://github.com/woowacourse/java-attendance/pull/49/commits/4a5f23a5b4a455358a0b907f6ce59313f865a086)

jinu0328 committed:
refactor: 출석 기록 출력시 view에서 날짜 기반 정렬 수행

sinb57 commented on src/main/java/domain/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,64 @@
+                LocalTime.parse(time, timeFormatter)));
+    }
+
+    private Crew getCrew(Crews crews, String nickname) {
+        try {
+            return crews.findByNickname(nickname);
+        } catch (IllegalArgumentException e) {
+            Crew crew = new Crew(nickname);
+            crews.add(crew);
+            return crew;
+        }
+    }
```
이 행위는 Crews가 지녀도 좋아보여요~ Map 자료구조의 put을 떠올려보아요

sinb57 reviewed (approved):
안녕하세요 에드~ 1단계 진행하느라 고생 많으셨어요!
리뷰 잘 반영해주셔서 1단계는 이만 머지할게요~ 2단계에서 봬요!!
